Permitir técnico se cadastrar com um clube existente

luc527 opened the pull request.

luc527 committed:
debounce and throttle utility functions

luc527 committed:
remove leftover test code

luc527 committed:
back-end da funcionalidade

Quando o técnico informar o nome de um clube que já existe no campo
"clube", o back-end irá vincular o técnico com esse clube em vez de
criar um clube novo com o mesmo nome. Ou seja, não é mais o front-end
que decide se o clube é novo ou não. No lugar dos campos mutualmente
exclusivos 'idClubeExistente' e 'nomeClubeNovo', existe somente um
campo 'clube' no payload enviado ao servidor.

Todos os testes foram atualizados para refletir isso.

Ainda podemos melhorar o front-end mostrando sugestõs de clubes
existentes conforme ele digita o nome de um clube. Quando ele selecionar
um clube da lista, o nome do clube será copiado ao input. Como o nome
enviado ao servidor será exatamente igual ao nome de um clube existente,
o técnico cadastrado será vinculado ao clube que selecionou.

luc527 committed:
front-end da funcionalidade + ajuste qtd. resultados pesquisa

Foi esse o jeito que eu encontrei de fazer sem ter que implementar um
autocomplete tipo esse na mão:
https://www.w3schools.com/howto/howto_js_autocomplete.asp

Conforme o usuário digita o nome do clube, mostramos os clubes com nomes
parecidos como opções tipo radio num card de "clubes encontrados"
abaixo. Quando o usuário clicar numa dessas opções, o nome do clube é
copiado ao campo de texto. Ou seja, ainda é só um campo de texto que
enviamos ao servidor, mas agora existe um auxílio para que o valor dele
seja exatamente o nome de um clube que já existe, para evitar que um
clube novo seja criado só porque o usuário digitou o nome um pouco
diferente.

Para os casos em que o clube do técnico ainda não está cadastrado no
sistema, existe uma opção (a primeira opção da lista) que é simplesmente
igual ao valor que está sendo digitado no campo de texto.

ghost commented:
![Passed](https://sonarqube.mateux.dev/static/communityBranchPlugin/checks/QualityGateBadge/passed.svg?sanitize=true)

# Analysis Details
## 4 Issues
- ![Bug](https://sonarqube.mateux.dev/static/communityBranchPlugin/common/bug.svg?sanitize=true) 0 Bugs
- ![Vulnerability](https://sonarqube.mateux.dev/static/communityBranchPlugin/common/vulnerability.svg?sanitize=true) 0 Vulnerabilities
- ![Code Smell](https://sonarqube.mateux.dev/static/communityBranchPlugin/common/code_smell.svg?sanitize=true) 4 Code Smells

## Coverage and Duplications
- ![Coverage](https://sonarqube.mateux.dev/static/communityBranchPlugin/checks/CoverageChart/100.svg?sanitize=true) 100.00% Coverage (100.00% Estimated after merge)
- ![Duplications](https://sonarqube.mateux.dev/static/communityBranchPlugin/checks/Duplications/3.svg?sanitize=true) 0.00% Duplicated Code (0.00% Estimated after merge)

**Project ID:** MateuxDotDev_badminton-manager_AYfNu-knGxuymmoUfIu7

[View in SonarQube](https://sonarqube.mateux.dev/dashboard?id=MateuxDotDev_badminton-manager_AYfNu-knGxuymmoUfIu7&pullRequest=26)

MateuxLucax commented on public/assets/js/debounce-throttle.js:
```diff
@@ -0,0 +1,17 @@
+function debounce(t, f) {
```
na branch que estou mexendo (https://github.com/MateuxDotDev/badminton-manager/pull/27) também utiliza esses esquemas. Porém também criei um arquivo de utilitários extras (para filtragem), no arquivo `public/assets/js/utils.js`.

Será que não seria bom criar um arquivo que contenha mais desses utilitários? Ou ir criando arquivos conforme a necessidade. Tipo ao invés do `utils.js` lá naquela branch teria um `search.js` ou algo assim?

luc527 replied:
Feito, criei um `utils.js` e movi as funções para lá. Pode ficar tudo num arquivo `utils.js` mesmo.

MateuxLucax commented on public/cadastro/acao.php:
```diff
@@ -40,3 +41,51 @@ function realizarCadastro(array $req): Response
+    $filtro = implode(
+        ' AND ',
+        array_map(
+            fn($termo) => "nome like '%" . addslashes($termo) . "%'",
```
Talvez fazer alguma limpeza para evitar SQLInjection nessa parte? Me parece manipulação de string bruta apenas

luc527 replied:
Feito, troquei por query preparada (`nome ILIKE ?`)

MateuxLucax commented on public/cadastro/index.php:
```diff
@@ -42,36 +39,83 @@
         <div class="mb-5">
-            <label class="form-label">Clube</label>
-            <input type="text" class="form-control" name="clube" required/>
+            <div class="mb-3">
```
Ele é case-sensitive nas buscas. Não seria bom se não fosse? (dependendo do esforço que desse, n importa muito também).

luc527 replied:
Corrigido com um `ILIKE` no lugar de `LIKE`

luc527 committed:
fix

MateuxLucax commented on public/cadastro/index.php:
```diff
@@ -42,36 +39,83 @@
-            <input type="text" class="form-control" name="clube" required/>
+            <div class="mb-3">
+                <label class="form-label">Clube</label>
+                <input type="text" class="form-control" name="clube" autocomplete="off"/>
```
https://getbootstrap.com/docs/5.2/forms/form-control/#datalists

Será que um datalist não seria melhor? Esse do radio me parece muito complicado.
Aí só fazer um controle para buscar clube conforme o que o cara escreveu. Se existir um com mesmo nome (case-insensitive), utilizar ele como clube.

luc527 replied:
É uma boa, e de fato isso que eu fiz com os radios ficou mais complicado do que eu gostaria. Se eu soubesse do datalist teria tentando fazer com ele. Mas acho melhor colocar essa refatoração como uma tarefa pra focar nas tarefas mais importantes agora, aí alguém faz depois se quiser. (Na prática provavelmente vai ficar assim mesmo)

ghost commented:
![Passed](https://sonarqube.mateux.dev/static/communityBranchPlugin/checks/QualityGateBadge/passed.svg?sanitize=true)

# Analysis Details
## 4 Issues
- ![Bug](https://sonarqube.mateux.dev/static/communityBranchPlugin/common/bug.svg?sanitize=true) 0 Bugs
- ![Vulnerability](https://sonarqube.mateux.dev/static/communityBranchPlugin/common/vulnerability.svg?sanitize=true) 0 Vulnerabilities
- ![Code Smell](https://sonarqube.mateux.dev/static/communityBranchPlugin/common/code_smell.svg?sanitize=true) 4 Code Smells

## Coverage and Duplications
- ![Coverage](https://sonarqube.mateux.dev/static/communityBranchPlugin/checks/CoverageChart/100.svg?sanitize=true) 100.00% Coverage (100.00% Estimated after merge)
- ![Duplications](https://sonarqube.mateux.dev/static/communityBranchPlugin/checks/Duplications/3.svg?sanitize=true) 0.00% Duplicated Code (0.00% Estimated after merge)

**Project ID:** MateuxDotDev_badminton-manager_AYfNu-knGxuymmoUfIu7

[View in SonarQube](https://sonarqube.mateux.dev/dashboard?id=MateuxDotDev_badminton-manager_AYfNu-knGxuymmoUfIu7&pullRequest=26)

MateuxLucax commented on src/Tecnico/TecnicoRepository.php:
```diff
@@ -99,20 +99,52 @@ public function getViaId(int $id): ?Tecnico
+    {
+        $pdo  = $this->pdo;
+
+        $sql  = "SELECT id FROM clube WHERE nome = :nome";
```
<img width="508" alt="image" src="https://github.com/MateuxDotDev/badminton-manager/assets/35275029/dc3d14b3-377b-417e-ac5f-0cf2056dfc42">

Adicionar uma forma de não ser case-sensitive

luc527 replied:
Feito, troquei por um `ILIKE`

luc527 committed:
correções conforme revisão

- trocados = e LIKE por ILIKE para match de clubes ser case insensitive
- query preparada para evitar SQL injection
  - achei que addlashes era suficiente, mas pesquisei e não era

luc527 commented on public/cadastro/acao.php:
```diff
@@ -58,28 +58,28 @@ function pesquisarClubes(array $req): Response
-           LIMIT 20
+           SELECT c.id, c.nome, COUNT(t.id) qtd_tecnicos
+             FROM clube c
+        LEFT JOIN tecnico t
```
Corrigi também um erro onde clubes sem atletas não eram mostrados (por causa do JOIN, corrigido com LEFT JOIN)

ghost commented:
![Passed](https://sonarqube.mateux.dev/static/communityBranchPlugin/checks/QualityGateBadge/passed.svg?sanitize=true)

# Analysis Details
## 4 Issues
- ![Bug](https://sonarqube.mateux.dev/static/communityBranchPlugin/common/bug.svg?sanitize=true) 0 Bugs
- ![Vulnerability](https://sonarqube.mateux.dev/static/communityBranchPlugin/common/vulnerability.svg?sanitize=true) 0 Vulnerabilities
- ![Code Smell](https://sonarqube.mateux.dev/static/communityBranchPlugin/common/code_smell.svg?sanitize=true) 4 Code Smells

## Coverage and Duplications
- ![Coverage](https://sonarqube.mateux.dev/static/communityBranchPlugin/checks/CoverageChart/100.svg?sanitize=true) 100.00% Coverage (100.00% Estimated after merge)
- ![Duplications](https://sonarqube.mateux.dev/static/communityBranchPlugin/checks/Duplications/3.svg?sanitize=true) 0.00% Duplicated Code (0.00% Estimated after merge)

**Project ID:** MateuxDotDev_badminton-manager_AYfNu-knGxuymmoUfIu7

[View in SonarQube](https://sonarqube.mateux.dev/dashboard?id=MateuxDotDev_badminton-manager_AYfNu-knGxuymmoUfIu7&pullRequest=26)

MateuxLucax committed:
refac: alterações para remover code smells

ghost commented:
![Passed](https://sonarqube.mateux.dev/static/communityBranchPlugin/checks/QualityGateBadge/passed.svg?sanitize=true)

# Analysis Details
## 0 Issues
- ![Bug](https://sonarqube.mateux.dev/static/communityBranchPlugin/common/bug.svg?sanitize=true) 0 Bugs
- ![Vulnerability](https://sonarqube.mateux.dev/static/communityBranchPlugin/common/vulnerability.svg?sanitize=true) 0 Vulnerabilities
- ![Code Smell](https://sonarqube.mateux.dev/static/communityBranchPlugin/common/code_smell.svg?sanitize=true) 0 Code Smells

## Coverage and Duplications
- ![Coverage](https://sonarqube.mateux.dev/static/communityBranchPlugin/checks/CoverageChart/100.svg?sanitize=true) 100.00% Coverage (100.00% Estimated after merge)
- ![Duplications](https://sonarqube.mateux.dev/static/communityBranchPlugin/checks/Duplications/3.svg?sanitize=true) 0.00% Duplicated Code (0.00% Estimated after merge)

**Project ID:** MateuxDotDev_badminton-manager_AYfNu-knGxuymmoUfIu7

[View in SonarQube](https://sonarqube.mateux.dev/dashboard?id=MateuxDotDev_badminton-manager_AYfNu-knGxuymmoUfIu7&pullRequest=26)
